File: examples/dubbing/python/create_a_dub_from_url.py
import os
from typing import Optional

from dotenv import load_dotenv
from dubbing_utils import download_dubbed_file, wait_for_dubbing_completion
from elevenlabs.client import ElevenLabs

# Load environment variables
load_dotenv()

# Retrieve the API key
ELEVENLABS_API_KEY = os.getenv("ELEVENLABS_API_KEY")
if not ELEVENLABS_API_KEY:
    raise ValueError(
        "ELEVENLABS_API_KEY environment variable not found. "
        "Please set the API key in your environment variables."
    )

client = ElevenLabs(api_key=ELEVENLABS_API_KEY)


def create_dub_from_url(
    source_url: str,
    source_language: str,
    target_language: str,
) -> Optional[str]:
    """
    Downloads a video from a URL, and creates a dubbed version in the target language.

    Args:
        source_url (str): The URL of the source video to dub. Can be a YouTube link, TikTok, X (Twitter) or a Vimeo link.
        source_language (str): The language of the source video.
        target_language (str): The target language to dub into.

    Returns:
        Optional[str]: The file path of the dubbed file or None if operation failed.
    """

    response = client.dubbing.dub_a_video_or_an_audio_file(
        source_url=source_url,
        target_lang=target_language,
        mode="automatic",
        source_lang=source_language,
        num_speakers=1,
        watermark=True,  # reduces the characters used
    )

    dubbing_id = response.dubbing_id
    if wait_for_dubbing_completion(dubbing_id):
        output_file_path = download_dubbed_file(dubbing_id, target_language)
        return output_file_path
    else:
        return None


if __name__ == "__main__":
    source_url = "https://www.youtube.com/watch?v=0EqSXDwTq6U"  # Charlie bit my finger
    source_language = "en"
    target_language = "fr"
    result = create_dub_from_url(source_url, source_language, target_language)
    if result:
        print("Dubbing was successful! File saved at:", result)
    else:
        print("Dubbing failed or timed out.")

File: examples/dubbing/node/createADubFromUrl.ts
import * as dotenv from "dotenv";
import { ElevenLabs } from "elevenlabs"; // Assume this is the ElevenLabs TypeScript import
import { downloadDubbedFile, waitForDubbingCompletion } from "./dubbingUtils"; // Assuming corresponding TypeScript functions

dotenv.config();

const ELEVENLABS_API_KEY = process.env.ELEVENLABS_API_KEY;
if (!ELEVENLABS_API_KEY) {
  throw new Error(
    "ELEVENLABS_API_KEY environment variable not found. " +
      "Please set the API key in your environment variables."
  );
}

// Assuming the class is correctly imported and the constructor accepts an object with an API key
const client = new ElevenLabs({ apiKey: ELEVENLABS_API_KEY });

async function createDubFromUrl(
  sourceUrl: string,
  sourceLanguage: string,
  targetLanguage: string
): Promise<string | null> {
  const response: any = await client.dubbing.dubAVideoOrAnAudioFile({
    // Assuming the return type to be `any`. Please replace with the correct return type.
    source_url: sourceUrl,
    target_lang: targetLanguage,
    mode: "automatic",
    source_lang: sourceLanguage,
    num_speakers: 1,
    watermark: true,
  });

  const dubbingId = response.dubbing_id;
  const dubbingCompleted: boolean = await waitForDubbingCompletion(dubbingId);
  if (dubbingCompleted) {
    const outputFilePath: string = await downloadDubbedFile(
      dubbingId,
      targetLanguage
    );
    return outputFilePath;
  } else {
    return null;
  }
}

(async () => {
  try {
    const sourceUrl: string = "https://www.youtube.com/watch?v=0EqSXDwTq6U"; // Charlie bit my finger
    const sourceLanguage: string = "en";
    const targetLanguage: string = "fr";
    const result: string | null = await createDubFromUrl(
      sourceUrl,
      sourceLanguage,
      targetLanguage
    );

    if (result) {
      console.log("Dubbing was successful! File saved at:", result);
    } else {
      console.log("Dubbing failed or timed out.");
    }
  } catch (error) {
    console.error("Error:", error);
  }
})();
